### PR TITLE
Init new annotation tool

### PR DESCRIPTION
**Pull request by Splines**
See feature overview in [this comment](https://github.com/MaMpf-HD/mampf/pull/512#issuecomment-1627722606) down below.

### CODE REVIEW

**Comment by fosterfarrell9** [comment]
Where do all these new files in the `cypress` folder come from? They seem unrelated to the PR.

**Comment by Frodo161** [comment]
This branch adds the feature to add **annotations** to the video timeline. They can be used for **giving feedback** to the teacher, for **getting help**, for publishing something directly in the **video comments** or - simply - to make **personal notes**.

By default students can create annotations on the video timeline. Click on the button next to the speed control to open a dialogue.

Bildschirmfoto vom 2023-07-09 09-17-27

Enter the text for your annotation (you can use Latex!), choose a color and select the category. You can publish the content directly as a comment and - if the teacher enabled this feature - make your annotation visible for the teacher. If the teacher entered an "emergency-link" (see below) and if you choose "content" as category and then "definition" the dialogue will give you a link for further help.

Markers will appear on the video timeline once you click on "Save". If you click on one of them their content will be shown in the interactive area on the right.

Bildschirmfoto vom 2023-07-09 09-16-24

To see the annotations of students (which are marked as "visible for teacher") the teacher has to activate the emergency button on the lecture edit page (comments).

Bildschirmfoto vom 2023-07-09 09-07-24

If it is (not) activated there is still the posibility to (de)activate it for each single video. By default the option "inherit from lecture" is selected.

Bildschirmfoto vom 2023-07-09 09-07-54

Go again to the lecture edit page (preferences this time). There you can enter links (or select a lecture) which is shown to the students as "emergency-link", if they have problems with understanding a definition.

Bildschirmfoto vom 2023-07-09 09-07-14

To see the feedback of the student, teachers now have the "feedback player" which can be accessed on the media edit page (where the video was uploaded). In this very simple player, the teacher sees all the annotations (that are marked as "visible for teacher") together with a heatmap. The teacher can (de)select the categories of annotations that are shown to him at the bottom of the video.

Mistake annotations are slightly bigger than other annotations and are not included in the calculation of the heatmap!

Bildschirmfoto vom 2023-07-09 15-48-10

File: config/locales/en.yml [hunk 1]
```diff
@@ -323,6 +323,53 @@ en:
...
+        Where's the problem?
+      definition: 'definition'
+      argument: 'argument'
+      strategy: 'proof strategy'
+      previous_annotation: 'previous annotation'
+      annotation_position: 'go to annotation'
+      edit_annotation: 'edit annotation'
+      close_annotation_area: 'close area'
+      next_annotation: 'next annotation'
+      further_help: 'You need further help? Then click on:'
```
Comment (Splines) [inline config/locales/en.yml:348]: If prof did not enter a link at all, you still get shown this `further_help` message, but without a link, which is not ideal. Maybe only show this message if a link has been entered by the prof?

File: config/locales/de.yml [hunk 2]
```diff
@@ -322,6 +322,54 @@ de:
...
+      category: 'Kategorie'
+      note: 'Notiz'
+      content: 'Inhalt'
+      mistake: 'Fehler'
+      presentation: 'Darstellung'
+      where_problem: >
+        Wo liegt das Problem?
+      definition: 'Definition'
+      argument: 'Argument'
+      strategy: 'Beweisstrategie'
```
Comment (Splines) [inline config/locales/de.yml:341]: As a user, I'm overwhelmed with all these category options to choose from, for example I ask myself questions such as:
- What does "presentation" mean?
- What does "argument" mean for "where's the problem"? "Argument" is such a broad word.
- Why do I get a "Where's the problem" to choose from, but then there's also the option "mistake", which does not give me the "Where's the problem?" options. What's the difference between a "content" mistake and a "mistake"?
- Why do I only get a help link when I click on "content", then on "definition". Why not also for the other options, like "proof strategy"? Why not for the high-level option "mistake"?

I feel like we definitely need some help texts for the users here. Or even restructure it a tiny bit, e.g. remove "mistake" as top-level option.

File: app/assets/stylesheets/thyme.scss [hunk 3]
```diff
@@ -341,3 +418,234 @@ figure {
...
+  #preview-toggle {
+    input:checked + .slider {
+      background-color: #0D6EFD;
+    }
+    .switch {
+      vertical-align: middle;
+    }
+  }
+  #annotation-modal-preview {
+    height: 543px;
```
Comment (Splines) [inline app/assets/stylesheets/thyme.scss:441]: Don't hardcode variables like that. Instead use relative positioning and specify how this element should behave with respect to other elements, e.g. "It should have the same height as this other element" (use case for `height: 100%`). Right now, for long notes, this is the reason why the preview is cut off here:

image

File: app/views/annotations/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    <br>
+
+    <!--- VISIBLE FOR TEACHER? - CHECKBOX -->
+    <% annotations_status = Medium.find_by(id: @medium_id).get_annotations_status %>
+    <% if annotations_status == 1 %>
+      <%= f.label :visible_for_teacher, class: "checkbox inline" do %>
+        <span, title="<%= t('admin.annotation.visible_for_teacher_tooltip') %>"><b>
+          <%= t('admin.annotation.visible_for_teacher') %> &nbsp
+        </b></span>
+        <%= f.check_box :visible_for_teacher %>
```
Comment (Splines) [inline app/views/annotations/_form.html.erb:114]: Checkbox not at the same height as label:

image

**Review by Splines — COMMENTED** [review]
Pretty cool to play around with the new feature. Some first remarks referring to the UI.

**Comment by Splines** [comment]
Closed in favor of #annotation #533.

**Pull request closed without merging**